src/main/java/games/strategy/engine/lobby/server/login/RsaAuthenticator.java

```diff
@@ -0,0 +1,153 @@
+package games.strategy.engine.lobby.server.login;
+
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import games.strategy.util.Util;
+
+/**
+ * A class which implements the TripleA-Lobby-Login authentication system using RSA encryption
+ * for passwords.
+ */
+public class RsaAuthenticator {
+  private static final Cache<String, PrivateKey> rsaKeyMap = buildCache();
+  private static final String RSA = "RSA";
+  private static final String RSA_ECB_OAEPP = RSA + "/ECB/OAEPPadding";
+
+  @VisibleForTesting
+  static final String ENCRYPTED_PASSWORD_KEY = "RSAPWD";
+
+  @VisibleForTesting
+  static final String RSA_PUBLIC_KEY = "RSAPUBLICKEY";
+
+  private static Cache<String, PrivateKey> buildCache() {
+    return CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build();
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  static boolean canProcessResponse(final Map<String, String> response) {
+    return response.get(ENCRYPTED_PASSWORD_KEY) != null;
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  public static boolean canProcessChallenge(final Map<String, String> challenge) {
+    return challenge.get(RSA_PUBLIC_KEY) != null;
+  }
+
+  /**
+   * Adds public key of a generated key-pair to the challenge map
+   * and stores the private key in a map.
+   */
+  static void appendPublicKey(final Map<String, String> challenge) {
+    challenge.put(RSA_PUBLIC_KEY, storeKeypair());
+  }
+
+  private static String storeKeypair() {
```

Comment:
Nit: Since the JDK has already established `KeyPair` as a compound word, it might be better for consistency to use that form everywhere (e.g. `storeKeyPair()`, `generateKeyPair()`, etc.).